RationaleMCP/0031/differences.md

```diff
@@ -341,3 +341,313 @@ end 'LineB';
+
+> _generic-element_ → ~~_import-clause_ | _extends-clause_ |~~ _normal-element_ | _parameter-equation_
+
+> _parameter-equation_ → **parameter** **equation** _guess-value_ **=** _expression_ _comment_
+
+> _guess-value_ → **guess** `[(]` _component-reference_ `[)]`
+
+(One can consider more general use of parameter equations in the future, but for now they are only used for guess value parameters.)
+
+For example:
+```
+Real 'x';
+parameter equation guess('x') = 1.5;
+Real 'y'; /* Parameter equation above does not leave public section. */
+```
+
+Similar to `pre('x')`, `guess('x')` acts as an independent variable in the initialization problem, and instead of providing a parameter equation, it can be determiend by an initial equation.  Since an initial equation cannot be modified after translation, this form is useful when the full Modelica modification of `start` was final.
+
+For example:
+```
+  Real 'x';
+initial equation
+  guess('x') - 1.5 = 0;
+```
+
+When a guess value for `'x'` is needed to solve an equation (or system of equations), this equation has an implicit dependency on `guess('x')`, and it is an error if this causes `guess('x')` to be solved in the same equation system as `'x'`.
+
+While a guess value parameter is allowed to appear in equations in the same ways as a normal parameter, Flat Modelica models originating from full Modelica are only expected to have `guess('x')` appearing in an initial equation in solved form, if appearing at all:
+```
+  Real 'x';
+initial equation
+  guess('x') = 1.5;
+```
+
+Default parameter equations for guess value parameters shall be added as needed to obtain a balanced initialization problem.  For example:
+```
+  Real 'x';
+initial equation
+  'x'^2 + 'x' = 1; /* Needs guess value for 'x' */
+```
+should be conceptually extended to:
+```
+  Real 'x';
+  parameter equation guess('x') = 0.0; /* Default guess value. */
+initial equation
+  'x'^2 + 'x' = 1; /* Needs guess value for 'x' */
+```
+
+The need for a default equation can also come from direct use of `guess('x')`:
+```
+  Real 'x';
+initial equation
+  'x' = guess('x');
+```
+
+#### Arrays and records
+
+For arrays, full Modelica modification of `start` with `each` will be described below (**TODO**).  Here is a simple example without `each`:
+```
+  Real[3] 'x';
+  parameter equation guess('x') = fill(1.5, 3);
+  Real[3] 'y';
+  Real[2] 'z';
+initial equation
+  guess('y') = fill(1.5, 3);
+  guess('z'[1]) = 1.5;
+  guess('z'[2]) = 1.5;
+```
+
+Records are similar to arrays:
+```
+  record 'R'
+    Real 'a';
+    Real 'b';
+  end 'R';
+  'R' 'x';
+  parameter equation guess('x') = R(1.1, 1.2);
+  'R' 'y';
+  'R' 'z';
+initial equation
+  guess('y') = R(1.1, 1.2);
+  guess('z'.'a') = 1.1;
+  guess('z'.'b') = 1.2;
+```
+
+Combining arrays and records:
+```
+  record 'R'
```

Comment:
Having worked on implementing a "flat" variant of Modelica; that has been used successfully for several years that in some cases contain records (including possibly heterogenous array of records; although I'm not sure how much that case has been used) this seems to create several problems - without solving any.

I think one of the general problems here is that we are trying to design a language without having the tooling. Obviously having the tooling has the risk of locking us into sub-optimal solutions, but without the constraint of implementing the tooling we don't discover all of the issues.

Reply:
I'm not sure I understand.  Are you suggesting that we should remove records from Flat Modelica?

Reply:
No, I'm suggesting that we keep records (in particular parameter records) - including record bindings.